### PR TITLE
Add second level cache in RegistryAuthService

### PR DESCRIPTION
**Pull request by munishchouhan**
This PR will add a second level cache for token in RegistryAuthService

### CODE REVIEW

**Comment by munishchouhan** [comment]
Tested in local
`12:48:01.204 [io-executor-thread-2] DEBUG i.s.w.auth.RegistryAuthServiceImpl - Authority token lookup for endpoint: 'RegistryAuthServiceImpl$CacheKey(image:hrma017/dev, auth:RegistryAuth(realm:https://auth.docker.io/token, service:registry.docker.io, type:Bearer, refreshable:true, endpoint:https://auth.docker.io/token?service=registry.docker.io), creds:RegistryCredentials[username=***; password=****])' => <TOKEN> [from store]`

File: src/main/groovy/io/seqera/wave/auth/RegistryAuthServiceImpl.groovy [hunk 1]
```diff
@@ -66,10 +69,25 @@ class RegistryAuthServiceImpl implements RegistryAuthService {
...
         @Override
         String load(CacheKey key) throws Exception {
-            return getToken0(key)
+            return getToken(key)
         }
     }
 
+    protected String getToken(CacheKey key){
+        // check if there's a record in the store cache (redis)
+        def result = tokenStore.get(key.toString())
```
Comment (pditommaso) [inline src/main/groovy/io/seqera/wave/auth/RegistryAuthServiceImpl.groovy:78]: Not sure about this

Comment (munishchouhan) [reply src/main/groovy/io/seqera/wave/auth/RegistryAuthServiceImpl.groovy:79]: I have create another method to create test for it

Comment (munishchouhan) [reply src/main/groovy/io/seqera/wave/auth/RegistryAuthServiceImpl.groovy:79]: you are right 
method executed twice
```
14:17:16.890 [io-executor-thread-2] DEBUG i.s.w.auth.RegistryAuthServiceImpl - Authority token lookup for cachekey: 'RegistryAuthServiceImpl$CacheKey(image:hrma017/dev, auth:RegistryAuth(realm:https://auth.docker.io/token, service:registry.docker.io, type:Bearer, refreshable:true, endpoint:https://auth.docker.io/token?service=registry.docker.io), creds:RegistryCredentials[username=****; password=dck****])' => <TOKEN> [from store]
14:17:17.473 [io-executor-thread-2] DEBUG i.s.w.auth.RegistryAuthServiceImpl - Authority token lookup for cachekey: 'RegistryAuthServiceImpl$CacheKey(image:hrma017/dev, auth:RegistryAuth(realm:https://auth.docker.io/token, service:registry.docker.io, type:Bearer, refreshable:true, endpoint:https://auth.docker.io/token?service=registry.docker.io), creds:RegistryCredentials[username=****; password=****])' => <TOKEN> [from store]
```

Comment (munishchouhan) [reply src/main/groovy/io/seqera/wave/auth/RegistryAuthServiceImpl.groovy:79]: in first run, its runs twice but in subsequent runs only runs one

Comment (munishchouhan) [reply src/main/groovy/io/seqera/wave/auth/RegistryAuthServiceImpl.groovy:79]: no even if i dont create getToken method, same thing happens, this is happens only when a new wave instance starts, so i think this is expected

Comment (pditommaso) [reply src/main/groovy/io/seqera/wave/auth/RegistryAuthServiceImpl.groovy:79]: Pushing a fix

Comment (pditommaso) [reply src/main/groovy/io/seqera/wave/auth/RegistryAuthServiceImpl.groovy]: This 49af9372 should solve, but I it will break some tests. please have a look

Comment (munishchouhan) [reply src/main/groovy/io/seqera/wave/auth/RegistryAuthServiceImpl.groovy]: sure, i will fix the tests

File: src/main/groovy/io/seqera/wave/auth/RegistryAuthServiceImpl.groovy [hunk 2]
```diff
@@ -55,21 +57,45 @@ class RegistryAuthServiceImpl implements RegistryAuthService {
...
+        result = getToken0(key)
+        log.debug "Authority token lookup for cachekey: '$key' => $result"
+        // save it in the store cache (redis)
+        tokenStore.put(stableKey, result)
+        return result
+    }
+
     private LoadingCache<CacheKey, String> cacheTokens = CacheBuilder<CacheKey, String>
                     .newBuilder()
                     .maximumSize(10_000)
```
[the remark this answers is not in the document]
Comment (munishchouhan) [reply src/main/groovy/io/seqera/wave/auth/RegistryAuthServiceImpl.groovy:101]: done

[the remark this answers is not in the document]
Comment (pditommaso) [reply src/main/groovy/io/seqera/wave/auth/RegistryAuthServiceImpl.groovy:101]: Cannot find it

[the remark this answers is not in the document]
Comment (munishchouhan) [reply src/main/groovy/io/seqera/wave/auth/RegistryAuthServiceImpl.groovy:101]: https://github.com/seqeralabs/wave/blob/54e0eff5740b4fe1983c4738c76841697fd6bd27/src/main/groovy/io/seqera/wave/auth/cache/RegistryTokenCacheStore.groovy#L46

[the remark this answers is not in the document]
Comment (munishchouhan) [reply src/main/groovy/io/seqera/wave/auth/RegistryAuthServiceImpl.groovy:101]: should i change the one in here to 1h too?
https://github.com/seqeralabs/wave/blob/54e0eff5740b4fe1983c4738c76841697fd6bd27/src/main/groovy/io/seqera/wave/auth/cache/RegistryAuthCacheStore.groovy#L47

[the remark this answers is not in the document]
Comment (munishchouhan) [reply src/main/groovy/io/seqera/wave/auth/RegistryAuthServiceImpl.groovy]: Added config class, `private LoadingCache<CacheKey, String> cacheTokens` is instance variable so need to move the initialization to a method

**Comment by munishchouhan** [comment]
tests failing because of `toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit`

**Comment by munishchouhan** [comment]
> weird, maybe credentials are not applied?

It should be github secrets

**Comment by pditommaso** [comment]
trigger a build into another branch to double check if the changes in this PR or a general problem

**Comment by munishchouhan** [comment]
> trigger a build into another branch to double check if the changes in this PR or a general problem

tested here, not failing anymore with rate-limit problem
https://github.com/seqeralabs/wave/actions/runs/10217586648

**Comment by pditommaso** [comment]
Should be ready. Willing to a test in dev ?

**Comment by munishchouhan** [comment]
> Should be ready. @munishchouhan Willing to a test in dev ?

ok sure

**Comment by munishchouhan** [comment]
Tested on dev successfully
`15:50:49.154 [io-executor-thread-5] DEBUG i.s.w.auth.RegistryAuthServiceImpl - Registry auth token for cachekey: 'RegistryAuthServiceImpl$CacheKey(image:hrma017/test, auth:RegistryAuth(realm:https://auth.docker.io/token, service:registry.docker.io, type:Bearer, refreshable:true, endpoint:https://auth.docker.io/token?service=registry.docker.io), creds:RegistryCredentials[username=hrma017; password=dck****])' [key-a326ff30888ba450] => <TOKEN> [from store]`

**Review by pditommaso — APPROVED** [review]

**Pull request merged**